### PR TITLE
[SPARK-45402][SQL][PYTHON] Add UDTF API for 'eval' and 'terminate' methods to consume previous 'analyze' result

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR adds a Python UDTF API for the `eval` and `terminate` methods to consume the previous `analyze` result.

This also works for subclasses of the `AnalyzeResult` class, allowing the UDTF to return custom state from `analyze` to be consumed later.

For example, we can now define a UDTF that perform complex initialization in the `analyze` method and then returns the result of that in the `terminate` method:

```
def MyUDTF(self):
    @dataclass
    class AnalyzeResultWithBuffer(AnalyzeResult):
        buffer: str

    @udtf
    class TestUDTF:
        def __init__(self, analyze_result):
            self._total = 0
	    self._buffer = do_complex_initialization(analyze_result.buffer)

        @staticmethod
        def analyze(argument, _):
            return AnalyzeResultWithBuffer(
                schema=StructType()
		    .add("total", IntegerType())
		    .add("buffer", StringType()),
                with_single_partition=True,
                buffer=argument.value,
            )

        def eval(self, argument, row: Row):
            self._total += 1

        def terminate(self):
            yield self._total, self._buffer

self.spark.udtf.register("my_ddtf", MyUDTF)
```

Then the results might look like:

```
sql(
    """
    WITH t AS (
      SELECT id FROM range(1, 21)
    )
    SELECT total, buffer
    FROM test_udtf("abc", TABLE(t))
    """
).collect()

> 20, "complex_initialization_result"
```

### Why are the changes needed?

In this way, the UDTF can perform potentially expensive initialization logic in the `analyze` method just once and result the result of such initialization rather than repeating the initialization in `eval`.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds new unit test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @ueshin @allisonwang-db @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Implementation seems fine from a cursory look, but let me defer to @allisonwang-db and @ueshin for the design.

File: python/pyspark/sql/udtf.py
Comment (ueshin) [inline python/pyspark/sql/udtf.py:123]: I guess we should distinguish an empty string and `None`.

File: python/pyspark/worker.py
Comment (ueshin) [inline python/pyspark/worker.py:799]: We should use some error class here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:310]: Why do we need this part?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:181]: I'm not sure if we should have the whole `PythonUDTFAnalyzeResult` here..

File: python/pyspark/sql/tests/test_udtf.py
Comment (allisonwang-db) [inline python/pyspark/sql/tests/test_udtf.py:2336]: We want to pass in some string values returned from the `analyze` method when initiating the UDTF. Should this be part of the `__init__`  method of the UDTF?

File: python/pyspark/sql/udtf.py
Comment (allisonwang-db) [inline python/pyspark/sql/udtf.py:110]: I wonder why it's a string type. Anything that can be pickled should be fine right?

File: python/pyspark/sql/udtf.py
Comment (allisonwang-db) [inline python/pyspark/sql/udtf.py:123]: Hmm I am not sure where this should be introduced. Currently, the `AnalyzeResult` represents some logical planning information such as output schema, ordering, and partitioning, and it's a bit confusing to me how to use this buffer (maybe we should change its name..)

**Review by allisonwang-db — COMMENTED** [review]
I am still thinking what's the best way to achieve this. It seems that we want to return some states from the `analyze` method that can be used to initialize the UDTF. 

A UDTF is serializable. The AnalyzeResult can even return a new UDTF:
AnalyzeResult(schema="...", udtf=MyNewUDTF) where the `__init__` method of the UDTF contains the computed states from the analyze method call. We can just give users control over how they want to change the UDTF after calling the `analyze`method.

File: python/pyspark/sql/tests/test_udtf.py
Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py]: This is done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala]: Update, I switched it back to set the schema separately per your request.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala
Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala]: I removed this.

File: python/pyspark/worker.py
Comment (dtenedor) [reply python/pyspark/worker.py]: This part is gone now.

**Comment by dtenedor** [comment]
Hi @allisonwang-db @ueshin thanks for your reviews, these were good comments, please look again! I think the new API is better now.

File: python/pyspark/sql/udtf.py
Comment (ueshin) [inline python/pyspark/sql/udtf.py:88]: Do we need this change?

File: python/pyspark/worker.py
Comment (ueshin) [inline python/pyspark/worker.py:710]: We should use a clone of the argument by `dataclasses.replace(pickled_analyze_result)` as now that it's not `frozen=True`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:261]: We don't need `Option` here as it will always be set?

File: python/pyspark/sql/udtf.py
Comment (dtenedor) [reply python/pyspark/sql/udtf.py:88]: I found it was necessary to allow UDTFs to create subclasses of `AnalyzeResult`. This is the way that we plan for these functions to compute custom state in the `analyze` method just once per function call and then propagate this information to future class instances to be consumed by the `eval` and `terminate` methods.

File: python/pyspark/worker.py
Comment (dtenedor) [reply python/pyspark/worker.py:710]: Sounds good, done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:261]: I find we need this since if the UDTF does not include an `analyze` method, this buffer is empty, and corresponding tests using these functions fail unless we check that this buffer is non-empty before using it.

**Review by dtenedor — COMMENTED** [review]
Thanks @ueshin for your comments! Please take another look.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala
Comment (ueshin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:291]: So it won't happen here as this is running `analyze` method.
I think `PythonUDTFAnalyzeResult.pickledAnalyzeResult` should be `Array[Byte]` and it should be wrapped with `Some` in `Analyzer`.

Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala]: Right, good point. I changed this to make the `pickedAnalyzeResult: Array[Byte]` required in the `PythonUDTFAnalyzeResult`, and optional in the `PythonUDTF` class itself. In the latter case, the field will be `None` if the UDTF was defined with a static output schema instead of implementing an `analyze` method.

**Review by dtenedor — COMMENTED** [review]
Thanks @ueshin for your review! I implemented the suggested change.

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

File: python/pyspark/sql/udtf.py
Comment (allisonwang-db) [reply python/pyspark/sql/udtf.py:88]: It would be really good to add this as a comment :)

File: sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala
Comment (allisonwang-db) [inline sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala:539]: Shall we remove this?

File: python/pyspark/worker.py
Comment (allisonwang-db) [inline python/pyspark/worker.py:713]: This means the UDTF handler does not accept an analyzeResult object in its `__init__` method? The try...except block will always be invoked when we call `contruct_udtf` which can be expensive. I wonder if we can only call this once.

**Review by dtenedor — COMMENTED** [review]
Hi @allisonwang-db thanks for your review, I responded to your comments and added more documentation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:294]: Shall we use `PythonWorkerUtils.readBytes` as https://github.com/apache/spark/pull/43321 is merged?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala:142]: ditto, `PythonWorkerUtils.writeBytes`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala:140]: nit: `foreach` instead of `map`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala
Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala]: Sure, this is done.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala
Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala]: Done.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala
Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala]: Done.

**Comment by ueshin** [comment]
The failed tests seem not related to this PR. Let me merge this now.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**
